crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -128,37 +129,95 @@ impl<S: States> NextState<S> {
     }
 }
 
+/// The previous state of [`State<S>`].
+///
+/// This is used internally by Bevy and generally not meant to be used directly.
+/// Precise semantics of data in this resource are not specified.
+#[derive(Resource, Default, Debug)]
+pub struct PrevState<S: States>(pub Option<S>);
+
 /// Run the enter schedule (if it exists) for the current state.
 pub fn run_enter_schedule<S: States>(world: &mut World) {
     world
         .try_run_schedule(OnEnter(world.resource::<State<S>>().0.clone()))
         .ok();
 }
 
-/// If a new state is queued in [`NextState<S>`], this system:
-/// - Takes the new state value from [`NextState<S>`] and updates [`State<S>`].
-/// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
-/// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
-/// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
+/// Label to order all [`OnEnter`], [`OnTransition`], and [`OnExit`] systems.
+///
+/// All `OnEnter` will run before all `OnTransition`,
+/// and all `OnTransition` will run before all `OnExit`.
+///
+/// The order of `OnEnter` systems relative to each other (and so on) is ambiguous by default.
+#[derive(SystemSet, Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum StateTransitionSet {
+    /// All [`OnEnter`] schedules are executed in this set.
+    OnEnter,
+    /// All [`OnTransition`] schedules are executed in this set.
+    OnTransition,
+    /// All [`OnExit`] schedules are executed in this set.
+    OnExit,
+}
+
+/// Run all state transition systems for the given state type.
+///
+/// This function is deprecated. It is not used internally by Bevy,
+/// and will be removed in a future version.
+#[deprecated(since = "0.13.0")]
 pub fn apply_state_transition<S: States>(world: &mut World) {
-    // We want to take the `NextState` resource,
-    // but only mark it as changed if it wasn't empty.
+    apply_state_transition_on_enter::<S>(world);
+    apply_state_transition_on_transition::<S>(world);
+    apply_state_transition_on_exit::<S>(world);
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnExit(exited_state)`] schedule, if it exists.
+pub fn apply_state_transition_on_exit<S: States>(world: &mut World) {
```

Comment:
In order to get the most consistent ordering, I think we should perform the state transition in a separate system before calling `OnExit`. This would counteract the stochastic execution order of the state transition systems by ensuring that every state has completed its transition by the time any of the `OnExit` systems run.

Reply:
Makes sense.